### PR TITLE
Attempt to document challenge types

### PR DESCRIPTION
**Pull request by schoen**
This still needs some editing but I'm putting it into a PR so that other people can see and work on it.

### CODE REVIEW

**Comment by schoen** [comment]
Fixes #4153.

**Comment by SwartzCr** [comment]
@pconrad-fb feel free to add comments

File: docs/challenges.rst [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+  program on your server listening to port 443
+
+* Your DNS records aren't valid
+  Try checking your DNS records with a tool like the DNSchecker at
+  http://www.dnsstuff.com/ to make sure there are no serious errors.
+  Sometimes a DNS error still allows your site to load in a web
+  browser, but prevents the certificate authority from issuing a
+  certificate.
+
+DNS-01 challenge:
```
Comment (zjs) [inline docs/challenges.rst]: Another possible issue with DNS-01 would be not waiting long enough for the DNS change to propagate.

**Comment by schoen** [comment]
I don't think that is a possible reason because Certbot checks the authoritative DNS servers directly, rather than relying on any cache.  Have you found an instance where someone has claimed this was a reason for a DNS-01 validation failure?

**Comment by zjs** [comment]
> Certbot checks the authoritative DNS servers directly

Perhaps I'm looking for the wrong strings, but I can't find any code in Certbot that queries DNS servers.

However, even if Certbot did query an authoritative DNS server directly, Certbot is not guaranteed to get the same view of the world as the ACME server; many providers (e.g., Google Cloud DNS) use Anycast to route requests to the nearest of a set of global-distributed eventually-consistent authoritative nodes.

Eventually, ACME may support some notion of retry (see, e.g., https://github.com/ietf-wg-acme/acme/pull/292, https://github.com/ietf-wg-acme/acme/pull/297) at which point it may be able to hide this complexity from users.

> Have you found an instance where someone has claimed this was a reason for a DNS-01 validation failure?

I'm not aware of user reports, but this is something I've experienced.

**Comment by SwartzCr** [comment]
merging so that @pconrad-fb can provide edits

**Pull request merged**